### ISSUE TITLE
doctype decl for standards compliant html5 (not: Quirks mode)

### ISSUE THREAD

**Issue by reuleaux** [shirok/Gauche-makiki#9]
Programming my first Makiki web application (thank you so much for Makiki!),
and I have created some pages now (with corresponding routes) in the style
of the examples, like so
```scheme
(define (tmpl :key (content '()))
  `(sxml
     (html (@ (lang "en"))
       (head
	 (meta (@ (charset "utf-8")))
	 (link (@
		 (type "text/css")
		 (rel "stylesheet")
		 (href "/main.css")
		 ))
	 )
       (body
	 (div (@ (id "left"))
           (div (@ (id "nav"))
             (a (@ (href "/contacts")
                  (class "navlnk"))
               "Contacts")
             (a (@ (href "/new")
                  (class "navlnk"))
               "New")
             (a (@ (href "/search")
                  (class "navlnk"))
               "Search")
             ))
	 (div (@ (id "main"))
	   ,content
           )
	 ))
     ))


(define-http-handler "/foo"
  (^[req app]
    (respond/ok req (tmpl :content '(span "hi")))
    ))
```
with some template tmpl i.e. (which could even be shorter). - Now the page `/foo `renders just fine in the browser, but the problem is:
when I (right-click-) inspect the page (or any element on the page for that matter) - in chromium in my case (but I assume the same holds for other browsers as well) ,
I see a small blue icon: "open issues to see 1 issue" - I do that, and there I see:

_Page layout may be unexpected due to Quirks Mode_

Well yes, indeed! - the doctype declaration is missing, this is not html5 after all!

The difference may be more clearly seen with these two routes/pages:

```scheme
(define-http-handler "/bar"
  (^[req app]
    (respond/ok req "<html><head></head><body><div><span>hi from bar</span></div><body></html>")
    ))

(define-http-handler "/baz"
  (^[req app]
    (respond/ok req "<!doctype html><html><head></head><body><div><span>hi from baz</span></div><body></html>")
    ))
```
`/bar `has the same issue as `/foo` above, whereas `/baz `renders without such issues - in (standards compliant) html5 i.e.

It would be nice, if somehow such a doctype declaration could be provided, 
as an option maybe? - as another possible parameter: `...(sxml (html5 ...))` maybe? - I leave this up to you, where / in which manner to best provide such a doctype decl.

In the meantime I have helped myself with this - this works at least (in standards
compliant html5):
```scheme
(define-http-handler "/banana"
  (^[req app]
    (let ([doc (tree->string (sxml:sxml->html `(html
						 (head
						   (link
						     (@
						       (type "css")
						       (rel "stylesheet")
						       (href "/..." )
						       )
						     )
						   )
						 (body
						   (span "hi from banana")
						   )
						 )
			       ))])
      (respond/ok req (format "<!doctype html>~%~a" doc))
      )))
```
Thank you so much in advance.
-Andreas

**Comment by shirok** [comment]
Good point.  The sxml library certainly have asymmetry between pasing (xml -> sxml) and serialization (sxml -> xml), notably, `DOCTYPE` is treated as meta-information during parsing, so it doesn't appear in the resulting S-expression; there's no place to put `DOCTYPE` "node" in sxml.  In serialization, `sxml:sxml->html` doesn't care `DOCTYPE`; it is supposed to be prepended by the caller.  That's how Oleg (the author of SXML spec and library) designed it.

It does pose an issue when you let Makiki render SXML tree.  I think Makiki should provide `DOCTYPE`.

**Comment by shirok** [comment]
Pushed the change that adds HTML5 doctype automatically when the content is `(sxml (html ...))`.

By the way, `respond/ok` is capable of simple template substitution.  I just realized it was not documented, though.  Check out tests/sxml-tmpl.scm.

**Comment by reuleaux** [comment]
OK, thanks you so much - for the doctype decl! (And I yet have to check out the template mechanism ...).
